### PR TITLE
LPD-12320 - FDS TS refactor & tests

### PR DESCRIPTION
**Pull request by juanjofgliferay**
Related task https://liferay.atlassian.net/browse/LPD-12320

As part of the testing tasks related to show and enformce Data Set Actions in the fragment I try to clean up the parts of the FDS component close to the actions. This mainly involve:
- JS to TS convertion 
- moving utilities to isolated files
- include basic tests 

The main idea is keeping the functionality as it was (test should verify it) and fix minor issues. Major refactors are needed, like unifying the `ActionDropdown` that is used in the different views or polishing actions types.

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+	itemData: any;
+	itemId: number | string;
+}
+
+export interface IItemsActions {
+	data?: IItemActionsData;
+	href?: string;
+	icon?: string;
+	id?: string | number;
+	items?: IItemsActions[];
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts:99]: On the fly generated property `items` https://github.com/liferay-frontend/liferay-portal/pull/3983/files#diff-bed70a46e4545a7d5c232541bea08157b92a65b58ed561e8b7f3bb229782b3a6R188

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/actionItems/formatActionURL.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/actionItems/formatActionURL.ts:1]: @dsanz Here we have how param interpolation work in actions and the [test](https://github.com/liferay-frontend/liferay-portal/pull/3983/files#diff-5e63b40daa8691b123e457c9992ae97929c00f0b540784c1b6a7e1fca08210c5R37) 
Also tested for the [QuickActions](https://github.com/liferay-frontend/liferay-portal/pull/3983/files#diff-8429be49e45d79a7dd5f5c31048b72465a49f6ac6f3927b81c481a8502f4e9a9R74)

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx [hunk 3]
```diff
@@ -3,64 +3,45 @@
...
 import React, {useContext, useState} from 'react';
 
-import FrontendDataSetContext from '../FrontendDataSetContext';
+import FrontendDataSetContext, {
+	IFrontendDataSetContext,
+} from '../FrontendDataSetContext';
 import {ACTION_ITEM_TARGETS} from '../utils/actionItems/constants';
-import {formatActionURL} from '../utils/index';
+import filterItemActions from '../utils/actionItems/filterItemActions';
+import formatActionURL from '../utils/actionItems/formatActionURL';
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx:15]: At some point, we should reshuffle those utils, to be in the optimal scope (`/actions/utils/*`), and not in global scope. But, it doesn't have to be in this PR.

File: modules/apps/frontend-data-set/frontend-data-set-web/test/utils/actionItems/filterItemActions.spec.ts
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/test/utils/actionItems/filterItemActions.spec.ts:1]: `*.spec.ts` is a Playwright convention. Most if not all all Jest files in DXP have the same name as util, `filterItemActions.ts` in this case.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx [hunk 4]
```diff
@@ -109,15 +98,15 @@ function Actions({actions, itemData, itemId, menuActive, onMenuActiveChange}) {
...
+		const doAction = ({defaultPrevented}: {defaultPrevented: boolean}) => {
 			if (target?.includes('modal')) {
 				event.preventDefault();
 
 				if (target === MODAL_PERMISSIONS) {
 					openPermissionsModal(url);
 				}
 				else {
-					openModal({
+					openModal!({
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx:109]: These non-null assertions are kinda awkward. Would errors go away if we add `openModal`, `highlightItems` etc to `IFrontendDataSetContext`?

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/QuickActions.tsx [hunk 5]
```diff
@@ -26,7 +25,7 @@ function QuickActions({actions, itemData, itemId, onClick}) {
 						}
 						key={action.data?.id || action.label}
 						monospaced={false}
-						onClick={(event) =>
+						onClick={(event: any) =>
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/QuickActions.tsx:28]: There are quite a few `any` types in this PR. These are not ideal, most if not all of these can be improved, but we can leave that for followups. Just the initial conversion to TS is a big enough step.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts:1]: I don't know why Github decided this was a completely new file. This confused me for way too long :sweat_smile:

**Review by markocikos — COMMENTED** [review]
@juanjofgliferay Awesome effort! :tada: Few minor nits and comments inline.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx (hunk 3 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx:15]: That's the idea, but there are still some utilities in the `utils/index.js` file that need to be extracted and also others that do not have a clear scope.

File: modules/apps/frontend-data-set/frontend-data-set-web/test/utils/actionItems/filterItemActions.spec.ts
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/test/utils/actionItems/filterItemActions.spec.ts:1]: Based on my research, there is no clear convention. I've found `*.test.(ts|tsx|js|jsx)` and `*.spec.(ts|tsx|js|jsx)` accross the portal :joy: 
[Objects-web](https://github.com/liferay/liferay-portal/tree/master/modules/apps/object/object-web/src/main/resources/META-INF/resources/js/tests) -> `*.spec.tsx` 
[Content-dashboard-web](https://github.com/liferay/liferay-portal/tree/master/modules/apps/content-dashboard/content-dashboard-web/test/js/components) `*.test.js`  `*.test.tsx`

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx (hunk 4 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx:109]: It is already defined there https://github.com/liferay-frontend/liferay-portal/pull/3983/files#diff-82c0e084703354bbeba91216bc400b1c309d56e1695ec96b6205bedec0488a4bR39

Could not find a better solution :disappointed:

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx:109]: All functions in `IFrontendDataSetContext` should almost certainly not be optional (`?`). It makes no sense, `openModal` will always be available. I am the one to blame. IIRC, my reasoning at the time was to put optional on everything, as the more permissive check, and figure out later on what must always be available. Not really the wisest line of reasoning in retrospective :sweat_smile:

File: modules/apps/frontend-data-set/frontend-data-set-web/test/utils/actionItems/filterItemActions.spec.ts
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/test/utils/actionItems/filterItemActions.spec.ts:1]: Yeah, consistency in file naming throughout DXP is not in a great place :disappointed: At least what we can do is keep tidy our own backyard, with [FI naming conventions](https://github.com/liferay/liferay-frontend-projects/blob/master/guidelines/general/file_names.md):
```
 Tests for file src/**/*.js  -->  test/**/*.js
```

Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/test/utils/actionItems/filterItemActions.spec.ts:1]: :ok_hand:

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx (hunk 4 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx:109]: I'll make most of the functions required and remove the non-null assertions.

Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx:109]: image

:shrug:

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.tsx:109]: `executeAsyncItemAction: () => {},` I suppose. It does make sense that compiler requires an assignment in `createContext`. We know that there in only one [setting of context](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js#L839-L886), but in theory there could be others that don't include functions.

**Review by markocikos — APPROVED** [review]
@juanjofgliferay LGTM! :sparkles:

**Comment by juanjofgliferay** [comment]
ci:forward

**Comment by juanjofgliferay** [comment]
ci:report:2537063285

**Comment by juanjofgliferay** [comment]
ci:forward

**Comment by juanjofgliferay** [comment]
ci:forward

**Comment by juanjofgliferay** [comment]
ci:report:2538075130

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
@juanjofgliferay based on https://github.com/liferay-frontend/liferay-portal/pull/3983#issuecomment-1941479106, there are no unique failures. This is safe to force forward.

**Comment by juanjofgliferay** [comment]
@markocikos I've seen that failures are related with the ones mentionend in https://liferay.slack.com/archives/C3JBR21HA/p1707829630692219. Just double check

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:report:2538613224

**Comment by juanjofgliferay** [comment]
Failures are not related based on https://github.com/liferay-frontend/liferay-portal/pull/3983#issuecomment-1943412421

**Comment by juanjofgliferay** [comment]
ci:forward:force

**Pull request closed without merging**